### ISSUE TITLE
Little more insights on CPU and Memory Usage Statistics

### ISSUE THREAD

**Issue by akdev05** [amir20/dozzle#1899]
### The aim of this feature is to view statistics options such as:
1. highest CPU-intensive container group
2. highest memory-consuming container group

For example, If I monitor 20 total instances in dozzle, out of which 10 instances are of "Type A" container, 5 are of "Type B" container and 5 are of "Type C" container, Then I need to view metrics of how much CPU does each type of container use.

### What I used as a workaround was as follows: 
1. Used this endpoint `api/events/stream` to get the necessary data
2. group those based on image name *(Considering same image would be used to create multiple instances)*
3. Derive the necessary statistics

The output I derived was something like this image below:

<img width="761" alt="image" src="https://user-images.githubusercontent.com/87078332/194699622-1ef62cf0-810b-4d58-954a-8c4ae674d673.png">

**Comment by amir20** [comment]
Hi @akdev05. That's a good idea. But I do think it's a little unique to your use case. In my experience, I haven't seen a lot of people running a lot of containers of same type. It's usually a variety of containers. 

However, I am working on revamping the whole CPU and memory dashboard. I am currently working #1896 which should be able to capture trend over the last XX minutes. 

The first step is to show trends and then get to some of the more unique use cases. I do want to redesign the homepage. Right now I am thinking of just a table. 

Let me know what you think of a more general use case and then maybe with the right controls it can also solve your problem.

**Comment by akdev05** [comment]
Yeah @amir20 I agree with you that it's a unique use case. But In more production scenarios, they might be useful, since many of them run and load-balance multiple containers of the same type. *(Even I was able to find a production issue yesterday in my organization, with the grouped data which I mentioned in the Picture before, as the CPU and memory usage suddenly got heavy)*

Hmm, how about an option in the dashboard to group stats for similar images? Based on the initial container details, If same image is being used by two or more containers, we could give users an option to group the CPU and Memory metrics according to the `image-name`. The `containers-changed` event can be listened to enable/disable this new option.

Maybe I could tinker with this feature after you merge #1896 , Please do let me know If this works out! Thanks for your response.

**Comment by amir20** [comment]
Sounds good. Will share when I have something. The UI is moving slow as I am doing some D3.js I haven't done before.

**Comment by akdev05** [comment]
Great, just now saw D3.js, seems to be a good choice for viewing trends. We can discuss after you merge #1896 , Will look out for any ideas until then.

**Comment by amir20** [comment]
<img width="548" alt="Screen Shot 2022-10-10 at 3 34 55 PM" src="https://user-images.githubusercontent.com/260667/194962253-af535af3-8106-447a-85de-18d0a0fcd0cd.png">

You can use `amir20/dozzle:pr-1896` to see some CPU usage. It wasn't that much code but I am focusing on the container page first. Maybe I can show all the charts on the home page if it doesn't melt the CPU.

**Comment by akdev05** [comment]
@amir20 Hope it wouldn't melt 🤣. But my aim is to consolidate all the instances of the group. Maybe if so many containers (or many groups of similar containers, to be precise) are there, then we can add an option to make users choose what they need to see. What do you think of this?

**Comment by akdev05** [comment]
The way you used those charts are neat and clean without disturbing UI. Maybe you may get many suggestions from the public after this is released.

**Comment by amir20** [comment]
There is a lot of infrastructure changes that need to be done first before I can do aggregated groups. For example, in that screenshot above, I am saving one stat per second. It's even more complicated because docker can send multiple stats per second so I have to throttle. If don't incorrectly, the browser can come to a halt.

**Comment by akdev05** [comment]
I would rather test your changes along with my desired changes by replicating my production scenario and let you know.

Also, more testing is needed before we proceed with such a feature. I don't want this tool to be CPU hogging, since its vision of lightweight part would get spoiled. So I will try to find some way to make this feature more CPU-friendly without hogging much usage.

Anyways, thanks for your immediate response.

**Comment by akdev05** [comment]
I saw your merged changes, and how you throttled the history. 
How about this idea? 

Since already you have adapted the `Container` model to get throttled data, I could create a component that gets such aggregated reduced data by throttling one more second? We could also limit the capacity *(since it's real-time and not a stored stat for analysis)* and it would be completely safe for CPU usage since data collection is throttled.

**Comment by akdev05** [comment]
Your suggestions on this please.

**Comment by amir20** [comment]
I am having a hard time understanding what you mean

> I could create a component that gets such aggregated reduced data by throttling one more second? We could also limit the capacity

What does that mean? 

I think my earlier point was that I have to be careful with how I design all the stats which is why I was thinking I would the aggregate _after_ I have everything working.

**Comment by akdev05** [comment]
I think there is a bit of a misunderstanding here.

What I meant was I would get the reduced data *(Total CPU/Memory usage by containers of same group)* by mapping and reducing only `memory-usage` and `cpu-usage` from the existing `activeContainers` data you have. Also, this reduced data would be taken so with the same `useThrottledRefHistory ` as you did for getting CPU trends.

These are just my initial thoughts. I was just getting to know what you were thinking of so that, we can get an initial idea of how things would work.

**Comment by akdev05** [comment]
Looks like my previous explanation was way too abstract. Hope I'm clear now. Please let me know.

**Comment by amir20** [comment]
Oh I see. That makes sense. But that’s still solving it for your specific use case only. 

I was actually thinking some thing different. The dashboard could be redesigned to have a table of all containers. Then there is an option to say group by name, image or whatever. Based on the table rows, a pie chart or whatever chart can be updated dynamically. 

This way it would work for your use case and any other use case. It would be kind of like data analytics tool which could be kind of cool. 

Of course this is a lot work. 🤔

**Comment by akdev05** [comment]
What you think is even better and more generic in terms of use cases:

> I was actually thinking something different. The dashboard could be redesigned to have a table of all containers. Then there is an option to say group by name, image or whatever. Based on the table rows, a pie chart or whatever chart can be updated dynamically.

Also, I could connect now on how this could melt the CPU if handled improperly 👀.

This could be an analytics tool as well and is really getting shape. Happy that you got this idea.

**Comment by amir20** [comment]
Update on this. I have merged and deployed first step of getting insights. 

<img width="529" alt="Screen Shot 2022-10-14 at 1 58 22 PM" src="https://user-images.githubusercontent.com/260667/195942931-c1fa19d1-54fb-4ee7-b6f5-37e3a6ebef67.png">

When I get more time, I want to create a table dashboard and possibly work on the comment above.

**Comment by akdev05** [comment]
@amir20 this looks neat and clean. Also I'm too looking out for some presenting ideas on the home screen as you said. Will share here as soon as they get some shape.
